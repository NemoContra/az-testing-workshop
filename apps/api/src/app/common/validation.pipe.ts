import { BadRequestException, Injectable, PipeTransform } from '@nestjs/common';
import { createFlightSchema } from '@az-testing-workshop/shared/util/api-models';
import { ZodSchema } from 'zod';

@Injectable()
export class ValidationPipe implements PipeTransform {
  constructor(private readonly schema: ZodSchema) {
  }
  transform(value: unknown): boolean {
    try {
      return this.schema.parse(value);
    } catch (error) {
      throw new BadRequestException(error);
    }
  }
}
