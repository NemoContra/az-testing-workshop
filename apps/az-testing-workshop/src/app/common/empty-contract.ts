import { Contract } from '@az-testing-workshop/shared/util/api-models';

export const emptyContract: Contract = {
  contractNumber: '',
  id: '',
  start: '',
  end: '',
  premium: 0,
  person: {
    firstname: '',
    lastname: '',
    dateOfBirth: '',
  },
};
