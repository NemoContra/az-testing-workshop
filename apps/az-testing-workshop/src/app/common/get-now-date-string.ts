const fillString = '0' as const;
const getYear = (now: Date) => `${now.getFullYear()}`;
const getMonth = (now: Date) =>
  (now.getMonth() + 1).toString().padStart(2, fillString);
const getDay = (now: Date) => now.getDate().toString().padStart(2, fillString);

export const getNowDateString = () => {
  const now = new Date();
  return [getYear(now), getMonth(now), getDay(now)].join('-');
};
