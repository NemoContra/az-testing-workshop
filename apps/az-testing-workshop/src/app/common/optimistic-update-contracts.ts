import { Contract } from '@az-testing-workshop/shared/util/api-models';

export const optimisticUpdateContracts = (
  contracts: Contract[] | undefined,
  updatedContract: Contract
) =>
  contracts?.map((contract) =>
    contract.id === updatedContract.id ? updatedContract : contract
  );
