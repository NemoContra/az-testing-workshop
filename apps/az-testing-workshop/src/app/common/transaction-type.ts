export const transactionTypes = ['AenderungNachname', 'Kuendigung'] as const;
export type TransactionsType = (typeof transactionTypes)[number];

export const transactionQueryParamKey = 'transaktion';
export const transactionQueryParam = <T extends TransactionsType>(
  transactionType: T
) =>
  ({
    [transactionQueryParamKey]: transactionType,
  } as const);
