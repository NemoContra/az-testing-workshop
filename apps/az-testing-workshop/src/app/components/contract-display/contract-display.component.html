<div class="two-col-grid">
  @if (contract.contractNumber) {
  <nx-data-display label="Vertragsnummer"
    >{{ contract.contractNumber }}
  </nx-data-display>
  } @if (contract.start) {
  <nx-data-display label="Vertragsbeginn"
    >{{ contract.start | date }}
  </nx-data-display>
  } @if (contract.end) {
  <nx-data-display label="Vertragsende"
    >{{ contract.end ? (contract.end | date) : '-' }}
  </nx-data-display>
  } @if (contract.person.firstname) {
  <nx-data-display label="Vorname"
    >{{ contract.person.firstname }}
  </nx-data-display>
  } @if (contract.person.lastname) {
  <nx-data-display label="Nachname"
    >{{ contract.person.lastname }}
  </nx-data-display>
  } @if (contract.person.dateOfBirth) {
  <nx-data-display label="Geburtsdatum"
    >{{ contract.person.dateOfBirth | date }}
  </nx-data-display>
  }
</div>
