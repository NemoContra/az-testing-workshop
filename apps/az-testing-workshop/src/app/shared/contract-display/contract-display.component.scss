.two-col-grid {
  margin-top: 16px;
  display: grid;
  gap: 2rem;
  grid-template-columns: 1fr 1fr;
}
