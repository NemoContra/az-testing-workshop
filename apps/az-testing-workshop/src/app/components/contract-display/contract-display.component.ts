import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { NxDataDisplayModule } from '@aposin/ng-aquila/data-display';
import { DatePipe } from '@angular/common';
import { emptyContract } from '../../common/empty-contract';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'contract-display',
  standalone: true,
  styleUrl: './contract-display.component.scss',
  templateUrl: './contract-display.component.html',
  imports: [NxDataDisplayModule, DatePipe],
})
export class ContractDisplayComponent {
  @Input({ required: true }) contract: Contract = emptyContract;
}
