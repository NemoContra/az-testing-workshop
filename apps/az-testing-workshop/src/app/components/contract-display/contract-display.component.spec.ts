import { createComponentFactory } from '@ngneat/spectator/jest';
import { ContractDisplayComponent } from './contract-display.component';
import {
  NxDataDisplayComponent,
  NxDataDisplayModule,
} from '@aposin/ng-aquila/data-display';
import { MockModule } from 'ng-mocks';
import { registerLocaleData } from '@angular/common';
import localeDe from '@angular/common/locales/de';
import { LOCALE_ID } from '@angular/core';
import { mockContracts } from '@az-testing-workshop/shared/util/mock-data';

registerLocaleData(localeDe);

describe('ContractDisplayComponent', () => {
  const createComponent = createComponentFactory({
    component: ContractDisplayComponent,
    providers: [{ provide: LOCALE_ID, useValue: 'de-DE' }],
    overrideComponents: [
      [
        ContractDisplayComponent,
        {
          remove: { imports: [NxDataDisplayModule] },
          add: { imports: [MockModule(NxDataDisplayModule)] },
        },
      ],
    ],
    detectChanges: false,
  });

  it('should show the correct fields of a provided contract', () => {
    const spectator = createComponent({
      props: {
        contract:   {
          id: '123456789',
          contractNumber: '1/2345678/9',
          start: '2024-01-01',
          end: '2056-01-31',
          premium: 42.42,
          person: {
            firstname: 'Homer',
            lastname: 'Simpson',
            dateOfBirth: '1961-05-16',
          },
        },
      },
    });

    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();

    const dataDisplays = spectator.queryAll(NxDataDisplayComponent);

    expect(dataDisplays.length).toEqual(6);

    expect(dataDisplays[0].label).toEqual('Vertragsnummer');
    expect(dataDisplays[0].size).toEqual(undefined);
    expect(dataDisplays[0].orientation).toEqual(undefined);

    expect(dataDisplays[1].label).toEqual('Vertragsbeginn');
    expect(dataDisplays[1].size).toEqual(undefined);
    expect(dataDisplays[1].orientation).toEqual(undefined);

    expect(dataDisplays[2].label).toEqual('Vertragsende');
    expect(dataDisplays[2].size).toEqual(undefined);
    expect(dataDisplays[2].orientation).toEqual(undefined);

    expect(dataDisplays[3].label).toEqual('Vorname');
    expect(dataDisplays[3].size).toEqual(undefined);
    expect(dataDisplays[3].orientation).toEqual(undefined);

    expect(dataDisplays[4].label).toEqual('Nachname');
    expect(dataDisplays[4].size).toEqual(undefined);
    expect(dataDisplays[4].orientation).toEqual(undefined);

    expect(dataDisplays[5].label).toEqual('Geburtsdatum');
    expect(dataDisplays[5].size).toEqual(undefined);
    expect(dataDisplays[5].orientation).toEqual(undefined);
  });

  it('should show the fields of an empty contract if no contract is provided', () => {
    const spectator = createComponent();

    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();

    const dataDisplays = spectator.queryAll(NxDataDisplayComponent);

    expect(dataDisplays.length).toEqual(0);
  });
});
