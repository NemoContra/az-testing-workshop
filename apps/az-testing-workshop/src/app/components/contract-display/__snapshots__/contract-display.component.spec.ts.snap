// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractDisplayComponent should show the correct fields of a provided contract 1`] = `
<contract-display
  contract={[Function Object]}
>
  <div
    class="two-col-grid"
  >
    <nx-data-display
      label="Vertragsnummer"
    >
      1/2345678/9
    </nx-data-display>
    <nx-data-display
      label="Vertragsbeginn"
    >
      01.01.2024
    </nx-data-display>
    <nx-data-display
      label="Vertragsende"
    >
      -
    </nx-data-display>
    <nx-data-display
      label="Vorname"
    >
      Homer
    </nx-data-display>
    <nx-data-display
      label="Nachname"
    >
      Simpson
    </nx-data-display>
    <nx-data-display
      label="Geburtsdatum"
    >
      16.05.1961
    </nx-data-display>
  </div>
</contract-display>
`;

exports[`ContractDisplayComponent should show the fields of an empty contract if no contract is provided 1`] = `
<contract-display
  contract={[Function Object]}
>
  <div
    class="two-col-grid"
  >
    <nx-data-display
      label="Vertragsnummer"
    >
    </nx-data-display>
    <nx-data-display
      label="Vertragsbeginn"
    >
    </nx-data-display>
    <nx-data-display
      label="Vertragsende"
    >
      -
    </nx-data-display>
    <nx-data-display
      label="Vorname"
    >
    </nx-data-display>
    <nx-data-display
      label="Nachname"
    >
    </nx-data-display>
    <nx-data-display
      label="Geburtsdatum"
    >
    </nx-data-display>
  </div>
</contract-display>
`;
